Fixes #1017 by cascading __call to mounted ControllerCollection

jstewmon opened the pull request:
all method calls that are handled by ControllerCollection::__call
will be recorded and replayed for a mounted ControllerCollection.

forward method calls to already mounted ControllerCollections
in the list of controllers.

include support for route-level middlewares using a priority
queue for the callbacks. insertion order is preserved by converting
the priority to an array and negative count as the subordinate
priority

thebeline commented on src/Silex/EventListener/MiddlewareListener.php:
```diff
@@ -75,7 +75,7 @@ public function onKernelResponse(FilterResponseEvent $event)
         }
 
-        foreach ((array) $route->getOption('_after_middlewares') as $callback) {
+        foreach ($route->getOption('_after_middlewares') ?: array() as $callback) {
```
Is this functionally different?  Or personal preference?  Please elaborate.

jstewmon replied:
SplPriorityQueue implements Iterator, but cannot be cast to array.

Prior to this patch, `$route->getOption('_before_middlewares')` could return `null` or `array`. After this patch, it can return `null` or `SplPriorityQueue`.

I changed it, so that in the `null` case, the expression results in `array()`, which is the same as it was with the cast. But in the `SplPriorityQueue` case, the expression results in `SplPriorityQueue`.

thebeline commented:
A few things:

1) There are a number of things going on in this PR, and I would like to see the two parts (cascading calls, and before/after priorities) in separate PRs, as they address separate issues.

2) On one hand I like the caching of the cascading calls, but this raises a few issues in and of it's self:
* If you are going to cache, the cached calls should be called on both `mount` as well as `match`.
* Caching the calls prevents garbage collection from ever releasing references in the cache.  This could be mitigated by clearing the calls cache on flush.

jstewmon commented:
> 1) There are a number of things going on in this PR, and I would like to see the two parts (cascading calls, and before/after priorities) in separate PRs, as they address separate issues.

If separate commits would get this merged, I'd separate them, but I don't really view them as separate issues b/c the problem is that ControllerCollection middleware do not follow the same semantics as Application middleware, and this commit fixes that by addressing the two pieces of that problem.

> 2) On one hand I like the caching of the cascading calls, but this raises a few issues in and of it's self:

> If you are going to cache, the cached calls should be called on both mount as well as match.

`match` already has a caching mechanism - `ControllerCollection->defaultRoute` - the defaultRoute is cloned and passed to new Controller instances created in `match`.

> Caching the calls prevents garbage collection from ever releasing references in the cache. This could be mitigated by clearing the calls cache on flush.

I think this is the case for all callbacks passed into a Silex method - nothing is ever unset to remove references to callbacks passed into any type.

stof commented:
there is a big issue with this change: handling 2 requests with your Silex app is not possible anymore, because the first request will empty the priority queue when iterating over it.

jstewmon commented:
@stof, thanks for your comment:

> there is a big issue with this change: handling 2 requests with your Silex app is not possible anymore, because the first request will empty the priority queue when iterating over it.

I wondered about this, but there are no test cases that demonstrate how it would come to pass that the list of Route middleware would be iterated multiple times. If you can provide a test case that illustrates the problem, I think there is a simple fix.

davedevelopment commented:
@jstewmon 

``` php
<?php

use Symfony\Component\HttpFoundation\Request;
use Silex\Application;

require "vendor/autoload.php";

$app = new Application();
$app['counter'] = 0;

$app->get('/', function () use ($app) {
    return $app['counter'];
})->before(function () use ($app) {
    $app['counter'] = $app['counter'] + 1;
});

for ($i = 0; $i < 5; $i++) {
    echo $app->handle(Request::create('/'))->getContent() . "\n";
}
```

Cloning the queue before it is iterated should fix it?

jstewmon commented:
@davedevelopment, thanks for the example.

Cloning the queue is exactly what I had in mind. I updated the PR branch to clone and added a test to demonstrate the use case (subsequent push was to fix an unrelated linting error in ApplicationTest.php).
